Fixes to find links (missing exceptions, exclude mailto: ...)

busykoala opened the pull request:
Close #39 
Close #37
Close #28 
Close #29 
Close #31 
Close #15 
Close #41

With this commit I fix unicode errors, missing exceptions and exclude
some links we do not want to report like mailto:.
Also an important change was to recursively yield through objects
searching for PloneSites, since they are not all in the same directory
sometimes.

Add logger and BrokenLink obj counter

Commit:
Introduce BrokenLink class initializing link information.

Commit:
Add functionality to BrokenLink class.

Commit:
Remodelling code making use of the BrokenLink class.

Commit:
Only have excel file in ram instead of saving into egg cache

With this change the excel workbook will not be saved physically
anymore but in ram only.
Also I added a better output to the excel sheet comparing test showing
the diff of the two dataframes loaded from the excel workbooks.

Commit:
Change config file loading

Commit:
Change config file loading

Commit:
Fixes to find links (missing exceptions, exclude mailto: ...)

With this commit I fix unicode errors, missing exceptions and exclude
some links we do not want to report like mailto:.
Also an important change was to recursively yield through objects
searching for PloneSites, since they are not all in the same directory
sometimes.

Commit:
Add logger and BrokenLink obj counter

This is to make working with the code easier.

Commit:
Fix: Handle paths having too many parent pointers

When a path has too many parent pointer we need to joint the path
with the root and platform again.

Commit:
Fix: implement command to start linkchecker correctly

Commit:
Fix: Do not call old function to get site objs

Previosly the function getting the plone site objs was named differently.
That lead to failure of five tests. Now the plone site objs can be found
.

Commit:
Fix: handle test having relations in textareas

Because in plone 4.x we have a basestring which always has a slash at
the end, relative paths work differently then expected by concatinating
them using urljoin.
This was the reason we the tests were still failing. With plone >= 5 we
need to look at that again since the basestring won't be usable anymore.

I also removed a test testing for /resolveuid which actually just ends
up in a broken relation and does not need to be tested separately.

Commit:
Add tests for link joining for relations from textareas.

Commit:
Use portal path instead of site id to get administrator email

The plone site id might not be an absolute identifier. Using the portal
path we can make sure that there is no redundancy.

Commit:
Add basic logger

To be able to log the events of the linkchecker one can now pass in a
pass in a path where a logfile is created.
Right now some basic information is logged.

Commit:
Add support for finding external links in non dexterity links.

The field content in non dexterity links does not have a href. Therefore
they were not found with the regex used. To avoid changing the methods
of the main use case I added a fake href around the link if it is a
basestring and has a valid scheme.

Commit:
Add support finding uids in archetype ReferenceFields

To find not only external Links but also relations this commit is
necessary.
Also I had to change the regex since it did only look for uids when
they started off in the string but not if the uid was somewhere in the
middle of a string.

Commit:
Add tests for archetype links and relations

Commit:
Remove portal path from link target/origin if there

Since the linkchecker does not have access to the information where the
pages run I remove the portal path so that the link can be copy/pasted
after the domain the page runs on.

Commit:
Add pages base uri in config to complete links in report

So that we can just click on a link in the report and do not need to
concatenate it with the pages domain I added the domain into the config.
The links in the report will be concatenated already and one can just
click to open the page.

Commit:
Add timeout config per site

maethu commented on ftw/linkchecker/command/broken_link.py:
```diff
@@ -0,0 +1,73 @@
+        self.header_location = 'Unknown header location'
+        self.error_message = 'No error occurred'
+        self.counter = BrokenLink.counter
+        BrokenLink.counter += 1
```
Do you need this for anything?
It's still not a good pattern to manipulate or use the class variable like that.

busykoala replied:
I was to add this information to the mail report like "n links were checked of which m were corrupt". Since I'm not using the information right now I'm removing it in this branch.

maethu commented on ftw/linkchecker/command/broken_link.py:
```diff
@@ -0,0 +1,73 @@
+            value = getattr(self, attr, '')
+            yield value if isinstance(value, basestring) else ''
+
+    def __len__(self):
```
Is this used somewhere?
I'm not sure what I expect when I do:

```
obj = BrokenLink()
print len(obj)
>>> 8
```

O think this is kinda misleading and using `len(obj.table_attrs)` would be totally fine and more readable.

busykoala replied:
Yes it is. But since the naming might be misleading I remove the length function from the class and use .table_attrs instead where I was using the length.

maethu commented on ftw/linkchecker/command/broken_link.py:
```diff
@@ -0,0 +1,73 @@
+    def __len__(self):
+        return len(self.table_attrs)
+
+    def complete_information_with_internal_path(self, obj_having_path, path):
```
Is it possible to set the status_code to 404 if the internal link is not working?

This way we could filter the Excel for all 404 links (internal and external) --> That's probably the main purpose.

busykoala replied:
I added a separate Issue for that (#43)

maethu commented on ftw/linkchecker/linkchecker.py:
```diff
@@ -3,80 +3,58 @@
-
-def work_through_urls(urls):
+def work_through_urls(external_link_objs, timeout_config):
+    global TIMEOUT
```
Instead of defining it like that you could just passt it to `get_uri_response`. 

setting a global variable in a function makes only an option if it's really necessary. 
Since it makes debugging and analysing the code unnecessary  more difficult. It could be set anywhere in that file

busykoala replied:
I'm gonna do that in #44 so we can proceed with this pull.

maethu replied:
👍

maethu commented on ftw/linkchecker/report_generating.py:
```diff
@@ -1,40 +1,76 @@
+LABELS.error_message = 'Error Message'
+LABELS = [LABELS]
+
+NUMBER_OF_LABELS = 8
```
That's `len(LABELS.table_attrs)` ? This way you can avoid that.

busykoala replied:
That's definitely the better way to do that. I'm adapting that.

maethu commented on ftw/linkchecker/test_setup.py:
```diff
@@ -5,6 +5,43 @@
 from zope.component.hooks import setSite
 
 
+def add_textarea_to_plone_site(portal):
```
`add_textarea_to_page`

busykoala replied:
👍

maethu reviewed (commented):
Just a few miner things you could fix. 

IMHO it does look good.

Commit:
Fix remove counter

Since we do not gain a lot of information out of the BrokenLink obj
counter I decided to remove it for now.

Commit:
Remove len from BrokenLink obj and use .table_attrs instead

The length was only used once and also misleading named. So I removed
the length function in the BrokenLink class and added the .table_attrs
where the length was used.

Commit:
Fix: Use length of table_attrs instead of a fixed value

To scale the number of cols the excel autofilter has I used a fixed
value. Since we have these attributes being in the excel sheet in the
list table_attrs we can use the length of that instead.

Commit:
Fix: Correct naming

busykoala commented:
@maethu I fixed or referenced the things requested. Could you look at it again please?

maethu reviewed (approved):
I like the progress!! Let's move on! ✨
